fix efficiency of LZ4_compress_HC_destSize()

Cyan4973 opened the pull request:
LZ4_compress_HC_destSize() had a tendency
to discard its last match when this match overflowed specified dstBuffer limit.
The impact is generally moderate,
but occasionally huge,
typically when this last match is very large
(such as compressing a bunch of zeroes).

Issue #784 fixed for both Chain and Opt implementations.

Added a unit test suggested by @remittor checking this topic.

Cyan4973 committed:
fix efficiency of LZ4_compress_HC_destSize()

LZ4_compress_HC_destSize() had a tendency
to discard its last match when this match overflowed specified dstBuffer limit.
The impact is generally moderate,
but occasionally huge,
typically when this last match is very large
(such as compressing a bunch of zeroes).

Issue #784 fixed for both Chain and Opt implementations.

Added a unit test suggested by @remittor checking this topic.

Cyan4973 committed:
fix incorrect counting

after truncation of last sequence

Cyan4973 committed:
ensure last match not too close to end

must respect MFLIMIT distance from oend

remittor commented:
Compare code:
https://github.com/lz4/lz4/blob/7a75b045bde89e77e3573f91821d0314251aac1d/lib/lz4.c#L1154-L1155
https://github.com/lz4/lz4/blob/89736e4e27358ed463c2482553cfae1b1176c797/lib/lz4hc.c#L731-L732

Src commit: https://github.com/lz4/lz4/pull/923/commits/e7fe105ac6ed02019d34731d2ba3aceb11b51bb1#diff-bcf1f5402173f298047c4fa808056f3bL719-R726

`RUN_MASK = 15`  => ` llAdd = (lastRunSize + 241) / 256;`

Why?

Cyan4973 commented:
Both formulaes are extremely close to each other,
and both are "correct" with regards to `lz4` guarantees,
aka they won't overwrite the destination buffer, 
and both try to pack as much as possible data into this space.

That being said, the new formula is _slightly_ more accurate, 
hence pack _a bit_ more data into remaining space,
though this is only visible on large or specific lengths.

Examples : 

| Space left after token | Old formula : Nb of Last Literals / Effective size | New formula : Nb of Last Literals / Effective size |
| --- | --- | --- |
| 5 | 5 / 5 | 5 / 5 |
| 14 | 14 / 14 | 14 / 14 |
| 15 | 14 / 14 | 14 / 14 |
| 16 | 15 / 16 | 15 / 16 |
| 269 | 268 / 269 | 268 / 269 |
| 270 | 268 / 269 | __269__  / 270 |
| 271 | 269 / 270 | 269 / 270 |
| 272 | 270 / 272 | 270 / 272 |
| 4094 | 4078 / 4094 | 4078 / 4094 |
| 4096 | 4079 / 4095 | __4080__  / 4096 |
| 4112 | 4095 / 4112 | 4095 / 4112 |
| 65536 | 65279 / 65535 | __65280__ / 65536 |
| 100000 | 99607 / 99998 | __99609__ / 100000 |

The old formula was a bit over-cautious.
It was calculating the amount of additional length bytes required to represent the last segment length,
and then subtracting that amount to produce the literal length. Thus, the literal length was actually a bit shorter,
and maybe could be represented using _less_ additional bytes.
Hence, it was a bit of an over-estimation, but felt "good enough" at the time.

The new formula is implemented into `lz4hc` because it was the center of this investigation. That being said, it can be extended to `lz4` as well.

Cyan4973 committed:
improved last literals run on LZ4_compress_destSize

applying new more accurate formula from LZ4_compress_HC_destSize()

also : fix some minor display issue in tests/frametest

remittor commented:
> That being said, the new formula is _slightly_ more accurate,
> hence pack _a bit_ more data into remaining space,

The new formula is not only more accurate, but also faster than the old one, since division by 256 is elementary.
I like the new version!
